firedrake/preconditioners/fdm.py

```diff
@@ -569,219 +648,162 @@ def set_values(self, A, Vrow, Vcol, addv, triu=False):
-            index = index.reshape((2, -1))
-            numpy.copyto(index, node_map.values_with_halo[e])
-            index[1] += node_map.offset
-
-    def copy_indices(self, e):
-        for index, node_map in zip(self.indices, self.node_maps):
-            numpy.copyto(index, node_map.values_with_halo[e])
-
-    def add_offsets(self):
-        for index, node_map in zip(self.indices, self.node_maps):
-            index += node_map.offset
-
-    def assemble(self, A, addv=None, triu=False):
-        if A.getType() == PETSc.Mat.Type.PREALLOCATOR:
-            kernel = lambda *args, result=None: result
-        else:
-            kernel = self.kernel
-        result = self.kernel.result
-        insert = self.setSubMatCSR(PETSc.COMM_SELF, triu=triu)
-
-        # Core assembly loop
-        if self.layers is None:
-            for e in range(self.nel):
-                self.set_indices(e)
-                insert(A, kernel(*self.kernel_args, result=result),
-                       self.map_rows(), self.map_cols(), addv)
-        else:
-            for e in range(self.nel):
-                self.set_indices(e)
-                for _ in range(self.layers[e]):
-                    insert(A, kernel(*self.kernel_args, result=result),
-                           self.map_rows(), self.map_cols(), addv)
-                    self.add_offsets()
+        if A.getType() == "preallocator":
+            args = assembler.arguments[:2]
+            kernel = ElementKernel(PETSc.Mat(), name="preallocate").kernel(mat_type=mat_type, on_diag=on_diag)
+            assembler = op2.ParLoop(kernel, Vrow.mesh().cell_set,
+                                    *(op2.PassthroughArg(op2.OpaqueType("Mat"), arg.data) for arg in args),
+                                    *indices_acc)
+        assembler.arguments[0].data = A.handle
+        assembler()
 
 
 class ElementKernel(object):
-    """
-    A constant element kernel
-    """
-    def __init__(self, A, *coefficients):
-        self.result = A
-        self.coefficients = coefficients
-        self.integral_type = "cell"
-
-    def __call__(self, *args, result=None):
-        return result or self.result
-
-    def __del__(self):
-        self.destroy()
+    code = """
+PetscErrorCode %(name)s(const Mat A, const Mat B, %(indices)s) {
+    PetscFunctionBeginUser;
+    PetscCall(MatSetValuesSparse(A, B, %(rows)s, %(cols)s, %(addv)d));
+    PetscFunctionReturn(PETSC_SUCCESS);
+}"""
 
-    def destroy(self):
-        pass
+    def __init__(self, A, name=None):
+        self.result = A
+        self.mats = [self.result]
+        self.name = name or type(self).__name__
+        self.rules = {}
+
+    def make_args(self, *mats):
+        return [op2.PassthroughArg(op2.OpaqueType(mat.klass), mat.handle) for mat in list(mats) + self.mats]
+
+    def kernel(self, mat_type="aij", on_diag=False, addv=None):
+        if addv is None:
+            addv = PETSc.InsertMode.INSERT
+        indices = ("rindices", "cindices")[:2-on_diag]
```

Comment:
```suggestion
        indices = ("rindices",) if on_diag else ("rindices", "cindices")
```
? I think is more readable.